### PR TITLE
[export_python] precompile: make capture leave the caller's RNG state untouched

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #193901
* #193900
* #193899
* #198257
* #198256
* #198255
* #198254
* #198253
* #198252
* #193898
* __->__ #198247
* #198246
* #198245
* #193974
* #193973

**Problem.** make_fx runs `fn` for real, so capturing a graph that draws (a
`torch.rand`, a dropout in training mode) advances the very generators the
caller's next real call draws from. Asking for an artifact thereby changed the
caller's random numbers, which is a side effect of compiling rather than of
running anything. The obvious fix, snapshot every generator and always write it
back, is wrong in a threaded process: the restore writes process-global state, so
rewinding when the capture drew nothing silently replays whatever a concurrent
thread drew meanwhile.

**Fix.** Under the capture lock, `_CaptureRngState` snapshots the CPU generator
plus every CUDA/XPU generator that is already initialized or reachable from the
arguments, reading them with all modes disabled so a tracing or torch-function
mode cannot intercept the reads. After capture, `_graph_rng_devices` inspects the
traced graph: an op tagged `nondeterministic_seeded` counts as a draw on its
output's device, except that the ops in `_RNG_GATED_BY_ARG` (dropout, attention,
RNN, rrelu) are excused when a literal argument such as `dropout_p=0.0` or
`train=False` proves they cannot draw, and any non-aten op makes the answer
"could be any generator", since an opaque custom op can draw inside its own
kernel. Only generators the graph could have drawn from are restored. A
fake-traced (mark_unbacked) capture runs no real kernel and restores nothing, a
capture that raises restores nothing (there is no graph to attribute draws to,
and capture rejections are routine), and two warnings cover what cannot be
undone: a draw on a device whose state was never saved, and generator state that
changed although no graph op accounts for it. The precompile docstring documents
these rules, including that attribution is by output device, so a draw through an
explicit `torch.Generator` restores that device's default generator instead.

```python
torch.manual_seed(0)
before = torch.random.get_rng_state()
torch.compiler.precompile(lambda a: torch.rand_like(a), torch.empty(4), backend="eager")
assert torch.equal(torch.random.get_rng_state(), before)  # used to have advanced
```

Test Plan:

```
python test/test_precompile.py
python test/test_precompile.py -k rng_state
```

Adds test_capture_leaves_the_ambient_rng_state_alone (CPU generator) and
test_capture_leaves_the_accelerator_rng_state_alone (CUDA generator, skipped
without CUDA); both fail before this change.

This commit was authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude review this PR

**Comment by bobrenjc93** [comment]
Response to the previous review. The PR had grown past the size a review can finish, so three of the fixes are split into their own PRs directly above it.

**1. A `PrecompileError` raised after the trace leaves RNG advanced.** Fixed in #198275. The restore now runs before the post-trace validators, so a capture rejected after tracing still leaves the caller's generators where they were. A test pins this.

**2. The opaque-op scan skips `HigherOrderOperator` nodes.** Fixed in b22970bdecf. The check is inverted and fails closed: any call target that is not an `aten::` `OpOverload`, HOPs included, makes the device set unknown, so every snapshotted generator is restored.

**3. MPS is silently excluded.** Fixed in b22970bdecf. The hardcoded `("cuda", "xpu")` is gone. The snapshot finds the current accelerator the way `torch.random.fork_rng` does, through `torch.accelerator.current_accelerator()` and `torch.get_device_module`.

**4. Dead reference to `test_rng_gate_table_matches_the_op_registry`.** Fixed in #198277. The gate table moved into its own PR together with a real `test_rng_gate_table_matches_the_op_registry`, which rebuilds the table from the schema registry. The comment now names `training` as well, and `_cudnn_attention_forward_no_dropout_inplace` is listed as never drawing, matching its flash sibling.

**5. `nn.Module` args nested in a pytree.** Fixed in b22970bdecf. Module parameters and buffers are collected wherever the module sits in the argument pytree.

**6. No test covers the logic this PR adds.** Fixed across the split. This PR now has tests for a non-drawing graph restoring nothing, a CUDA-only draw leaving the CPU generator alone, and an opaque op restoring every generator. The gate-table tests (`p=0.0` / `dropout_p=0.0` leave the RNG untouched) are in #198277, and the raise-path test is in #198275. The CUDA tests use `skipUnless(TEST_CUDA, ...)` to match the file.

**7. The `drawn is None` warning fires when nothing is un-restorable.** Fixed in b22970bdecf. The devices are enumerated again after capture, and the warning fires only for a generator that was first initialized during capture and so was never snapshotted.

**8. An explicit `torch.Generator` rewinds the default generator.** Fixed in #198276. A draw that passes `generator=` no longer counts against its output device's default generator. The explicit generator can't be saved before capture names it, so it is left advanced with a warning, and the docstring says so.
